frame/offences/src/lib.rs

```diff
@@ -118,17 +139,31 @@ where
 		let slash_perbill: Vec<_> = (0..concurrent_offenders.len())
 			.map(|_| new_fraction.clone()).collect();
 
-		T::OnOffenceHandler::on_offence(
-			&concurrent_offenders,
-			&slash_perbill,
-			offence.session_index(),
-		);
+		Self::report_or_store_offence(&concurrent_offenders, &slash_perbill, offence.session_index());
 
 		Ok(())
 	}
 }
 
 impl<T: Trait> Module<T> {
+	/// Tries (without checking) report an offence. Stores them in [`DeferredOffences`] in case it
+	/// fails
+	fn report_or_store_offence(
+		concurrent_offenders: &[OffenceDetails<T::AccountId, T::IdentificationTuple>],
+		slash_perbill: &[Perbill],
+		session_index: SessionIndex,
+	) {
+		let _ = T::OnOffenceHandler::on_offence(
+			&concurrent_offenders,
+			&slash_perbill,
+			session_index,
+		).map_err(|_| {
+			<DeferredOffences<T>>::mutate(|d|
+				d.push((concurrent_offenders.to_vec(), slash_perbill.to_vec(), session_index))
```

Comment:
This is unbounded. Are there any guarantees that it won't go over some reasonable size? Also should we limit that?

Reply:
If a sane module implement `OnOffence`, since we check on every `on_initialise`, then it should be emptied pretty quickly. Yet, no we don't enforce any size. I would be happy to do that from a code PoV. But with slashing, I think it is a tricky situation. A bug here that allows someone that can report but not get slashed due to some buffer being full, IMO, is more catastrophic than temporarily storing lots of data.